File: SegmentTree/segment_tree.py
class SegmentTreeNode:
    def __init__(self, L, R, sum=0):
        self.L = L
        self.R = R
        self.left = None
        self.right = None
        self.sum = sum


class SegmentTree:

    def __init__(self, nums):
        self.nums = nums
        self.root = self.build_tree(0, len(nums) - 1)

    def build_tree(self, L, R):
        if (L == R):
            return SegmentTreeNode(L, R, self.nums[L])
        M = (L + R) // 2
        curr = SegmentTreeNode(L, R)
        curr.left = self.build_tree(L, M)
        curr.right = self.build_tree(M + 1, R)
        curr.sum = curr.left.sum + curr.right.sum

        return curr

    def update(self, index: int, val: int) -> None:
        self.update_for(self.root, index, val)

    def update_for(self, root, index, val):
        if root.L == root.R:
            root.sum = val
            return

        M = (root.L + root.R) // 2
        if index <= M:
            self.update_for(root.left, index, val)
        else:
            self.update_for(root.right, index, val)

        root.sum = root.left.sum + root.right.sum

    def query(self, L: int, R: int) -> int:
        return self.query_for(self.root, L, R)

    def query_for(self, root, L, R):
        if root.L == L and root.R == R:
            return root.sum

        M = (root.L + root.R) // 2
        if R <= M:
            return self.query_for(root.left, L, R)
        elif L > M:
            return self.query_for(root.right, L, R)
        else:
            return self.query_for(root.left, L, M) + self.query_for(root.right, M + 1, R)

File: SegmentTree/SegmentTree.java
package SegmentTree;

class SegmentTree {

    private SegmentTreeNode root = null;

    public SegmentTree(int[] nums) {
        root = buildTree(nums, 0, nums.length -1);
    }

    private SegmentTreeNode buildTree(int[] nums, int start, int end){
        if (start > end){
            return null;
        }

        SegmentTreeNode curr = new SegmentTreeNode(start, end);
        if (start == end) {
            curr.sum = nums[start];
        } else{
            int mid = start + (end - start) / 2;
            curr.left = buildTree(nums, start, mid);
            curr.right = buildTree(nums, mid + 1, end);
            curr.sum = curr.left.sum + curr.right.sum;
        }

        return curr;
    }
    public void update(int index, int val) {
        update(root, index, val);

    }

    private void update(SegmentTreeNode root, int index, int value){
        if (root.start == root.end){
            root.sum = value;
            return;
        }
        int mid = root.start + (root.end - root.start) /2 ;
        if (index <= mid){
            update(root.left, index, value);
        } else {
            update(root.right, index, value);
        }
        root.sum = root.left.sum + root.right.sum;
    }

    public int query(int i, int j) {
        return query(root, i, j);
    }

    private int query(SegmentTreeNode root, int start, int end){
        if(root.end == end && root.start == start){
            return root.sum;
        }

        int mid = root.start + (root.end - root.start) /2;

        if(mid >= end){
            return query(root.left, start, end);
        } else if(mid < start){
            return query(root.right, start, end);
        } else {
            return query(root.left, start, mid) + query(root.right, mid + 1, end);
        }
    }
}
